packages/models-library/tests/test_utils_string_substitution.py

```diff
@@ -0,0 +1,174 @@
+# pylint: disable=redefined-outer-name
+# pylint: disable=unused-argument
+# pylint: disable=unused-variable
+# pylint: disable=too-many-arguments
+
+import sys
+from pathlib import Path
+
+import pytest
+from models_library.utils.string_substitution import (
+    SubstitutionsDict,
+    TemplateText,
+    substitute_all_legacy_identifiers,
+    upgrade_identifier,
+)
+
+
+@pytest.mark.parametrize(
+    "legacy,expected",
+    [
+        (
+            "%%container_name.sym-server%%",
+            "OSPARC_ENVIRONMENT_CONTAINER_NAME_SYM_SERVER",
+        ),
+        (
+            "%service_uuid%",
+            "OSPARC_ENVIRONMENT_SERVICE_UUID",
+        ),
+    ],
+)
+def test_upgrade_identifiers(legacy: str, expected: str):
+    assert upgrade_identifier(legacy) == expected
+
+
+def test_substitution_with_new_and_legacy_identifiers():
+
+    stringified_config = """
+    compose_spec:
+        service-one:
+            init: true
+            image: ${SIMCORE_REGISTRY}/simcore/services/dynamic/dy-static-file-server-dynamic-sidecar-compose-spec:${SERVICE_VERSION}
+            environment:
+                # legacy examples
+                - SYM_SERVER_HOSTNAME=%%container_name.sym-server%%
+                - APP_HOSTNAME=%%container_name.dsistudio-app%%
+                - APP_HOSTNAME=some-prefix_%service_uuid%
+                - SPEAG_LICENSE_FILE=${OSPARC_ENVIRONMENT_SPEAG_LICENSE_FILE}
+                - MY_PRODUCT=$OSPARC_ENVIRONMENT_CURRENT_PRODUCT
+                - MY_EMAIL=$OSPARC_ENVIRONMENT_USER_EMAIL
+                - S4L_LITE_PRODUCT=yes
+                - AS_VOILA=1
+                - DISPLAY1=$${KEEP_DISPLAY}
+                - DISPLAY2=${SKIP_DISPLAY}
```

Comment:
I find it a bit confusing these SKIP_DISPLAY and KEEP_DISPLAY, not sure to what the result is. Maybe something like KEEP_DOLLAR_ESCAPE_SIGN would be more helpful.